xarray/tests/test_groupby.py

```diff
@@ -2013,29 +2022,29 @@ def test_upsample_nd(self):
         )
         assert_identical(expected, actual)
 
-    def test_upsample_tolerance(self):
+    def test_upsample_tolerance(self) -> None:
         # Test tolerance keyword for upsample methods bfill, pad, nearest
         times = pd.date_range("2000-01-01", freq="1D", periods=2)
         times_upsampled = pd.date_range("2000-01-01", freq="6h", periods=5)
         array = DataArray(np.arange(2), [("time", times)])
 
         # Forward fill
-        actual = array.resample(time="6h").ffill(tolerance="12h")
+        actual = array.resample(time="6h").ffill(tolerance="12h")  # type: ignore[arg-type] # TODO: tolerance also allows strings, same issue in .reindex.
         expected = DataArray([0.0, 0.0, 0.0, np.nan, 1.0], [("time", times_upsampled)])
         assert_identical(expected, actual)
 
         # Backward fill
-        actual = array.resample(time="6h").bfill(tolerance="12h")
+        actual = array.resample(time="6h").bfill(tolerance="12h")  # type: ignore[arg-type] # TODO: tolerance also allows strings, same issue in .reindex.
         expected = DataArray([0.0, np.nan, 1.0, 1.0, 1.0], [("time", times_upsampled)])
         assert_identical(expected, actual)
 
         # Nearest
-        actual = array.resample(time="6h").nearest(tolerance="6h")
+        actual = array.resample(time="6h").nearest(tolerance="6h")  # type: ignore[arg-type] # TODO: tolerance also allows strings, same issue in .reindex.
```

Comment:
Mypy is correct here. xarray has wrong typing. 

`tolerance` is supposed to be floats but obviously strings are also allowed according to this test. The rabbit hole goes deep here though, at least down to `.reindex`.

Reply:
tolerance can be string for DatetimeIndex.

Reply:
#8892